### PR TITLE
SVG icon support

### LINKED ISSUE #150: Support SVG assets
Supporting svg files in themes and adjusting them to the right size by pre-defined sizes in a global css would be nice.

### PR DESCRIPTION
**Pull request by ajmeese7**
This PR allows themes to include icons of type `SVG`, `PNG`, and `GIF`, the three primary formats used for images with transparency. The code makes a request to check if the default image type `.png` exists first, and if not it then checks for `.svg` and `.gif` images, respectfully. The styles in `_search.scss` were updated to enforce a standard size on `SVG` files, which could show up much larger than expected otherwise.

I chose this method because the proposed solutions in #150 were either:
1. Complicated and involved a lot of JSON configuration, which is an unnecessary barrier for users interested in introducing new themes
2. Didn't allow for mix-and-match of icon formats, which is something that is worth considering (I have had to introduce a `.png` in a theme because there was no equivalent `.svg` in the theme I was modifying)

My solution is not perfect, but it functions as intended. Reviews and critiques are of course always welcome :)

---

Closes #150
Refs #172

### CODE REVIEW

**Comment by ajmeese7** [comment]
I do not believe the failing unit tests have anything to do with the code that I introduced. However I can break the code out of the nested function into a separate function for CodeClimate if desired, that wouldn't be a problem.

File: src/utils/desktop.js [hunk 1]
```diff
@@ -110,6 +111,21 @@ export const isVisible = w => w &&
   !w.getState('minimized') &&
   w.getState('focused');
 
+/**
+ * Allows for requesting whether a file exists without `async`
+ * @todo Move to VFS, so the 404 error doesn't log to console
+ * @param {string} urlToFile
+ * @returns {boolean}
+ */
+const syncRequest = (urlToFile) => {
```
Comment (andersevenrud) [inline src/utils/desktop.js:120]: Synchronous requests blocks the main thread. So any time this function runs, the entire UI is frozen until the request finishes.

Also this use is deprecated in all browsers.

So there needs to be some alternate way of solving this. Ideally without any network request at all. The client should be aware of what's possible or not, and not do queries.

PS: It's possible to build OS.js in a standalone mode that does not require a server for use. And this would break those kinds of deployments as you can't do HTTP requests in a `file:///` environment.

**Review by andersevenrud — CHANGES_REQUESTED** [review]
I'm not really sure the solution suggested here is the optimal one.

See my annotation of code about the lookups.

> Complicated and involved a lot of JSON configuration, which is an unnecessary barrier for users interested in introducing new themes

>Didn't allow for mix-and-match of icon formats, which is something that is worth considering (I have had to introduce a .png in a theme because there was no equivalent .svg in the theme I was modifying)

Both of these can be solved if theme packages had some kind of way to automate generation of some kind of metadata so all icons are known (including their extensions).

Maybe with a simple webpack plugin supplied in dev-meta ? It's possible to get a list of every single file that went into a bundle.

**Comment by andersevenrud** [comment]
> I do not believe the failing unit tests have anything to do with the code that I introduced. 

Huh. It fails on tests that don't use any network interactions, so that's weird. Haven't seen this. Tried to re-run, but same result.

Going to try again in a while in case it's just something on Github's end. All of the new commits in this repo went through just fine.

**Comment by andersevenrud** [comment]
> Huh. It fails on tests that don't use any network interactions

Actually, heh, it does. A `icon()` call is made in there :)

> I do not believe the failing unit tests have anything to do with the code that I introduced

So the chances are that's the case after all.

**Comment by andersevenrud** [comment]
> Maybe with a simple webpack plugin supplied in dev-meta ? It's possible to get a list of every single file that went into a bundle. If list is not available, fall back to png.

An alternative would be just to provide a simple CLI utility where you gave a directory and it would update the metadata file with everything it found :man_shrugging:

**Comment by andersevenrud** [comment]
NB: Updated your PR description and corrected the issue reference syntax so that when this goes through things gets closed as intended.

**Comment by ajmeese7** [comment]
> Maybe with a simple webpack plugin supplied in dev-meta ? It's possible to get a list of every single file that went into a bundle. If list is not available, fall back to png.

Good idea! Any chance you have an example handy of how you would accomplish this / could give some ideas of how to create and implement?

> An alternative would be just to provide a simple CLI (in dev-meta ideally since you don't usually have osjs-cli in your packages) utility where you gave a directory and it would update the metadata file with everything it found

Same question for this suggestion :)

**Comment by andersevenrud** [comment]
> Good idea! Any chance you have an example handy of how you would accomplish this / could give some ideas of how to create and implement?

Probably find what you need here (except for things that changed between webpack 4 and 5):

https://webpack.js.org/contribute/writing-a-plugin/

But the gist is, tap into the compilation, get the list of assets, figure out which ones are the icons (usually pretty easy because of path patterns).

> Same question for this suggestion :)

Just do a glob (and then same as above), really. Then write it to the metadata file.

**Comment by andersevenrud** [comment]
As for providing this stuff:

It could just be a  `import plugin from '@osjs/dev-meta/themePlugin` or something like that.

The simpler script method could just be stored in a `src` file, then add it as `bin` in `package.json` . Then user can run `npx generateThemeMetadata <path>` or something like that.

**Comment by ajmeese7** [comment]
@andersevenrud the metadata file is not included in the `dist`, but I just wrote a `webpack` plugin that generates a JSON file (`iconTypes.json` unless specified otherwise) mapping the image name to the file type like the following:

```json
{
  "mail-reply-all": "svg",
  "mail-reply-sender": "gif",
  "mail-send": "png",
  "media-eject": "svg",
  ...
}
```

How to you recommend reading from the JSON file in the client to determine what image type to attempt to load? Do you just want all `icon()` function calls to have `await` and the function converted to `async` with `fetch`, or do you have a better idea? Not sure if I can `import` the JSON file or if it has to be requested.

**Comment by andersevenrud** [comment]
> the metadata file is not included in the dist

Correct. Not in dist, but it's read when you do package discovery so that the client gets access to whatever is in the file :)

**Comment by andersevenrud** [comment]
> but it's read when you do package discovery so that the client gets access to whatever is in the file :)

Just to elaborate a little bit here. When you do package discovery, all packages with the `osjs` stuff in `package.json` is picked up and stored into an array of paths.

These paths are then used to look up a `metadata.json` file in the same path.

All of the metadata files are then merged together, and client reads it when it starts up.

**Comment by ajmeese7** [comment]
> Correct. Not in dist, but it's read when you do package discovery so that the client gets access to whatever is in the file :)

Like I said, I went with the `webpack` plugin route that you suggested, not the `cli` method. Manipulating `manifest.json` is not an option with that route, since it is not one of the assets that `webpack` is copying to the `dist` folder.

That's why I asked the second portion:

> How to you recommend reading from the JSON file in the client to determine what image type to attempt to load? Do you just want all icon() function calls to have await and the function converted to async with fetch, or do you have a better idea? Not sure if I can import the JSON file or if it has to be requested.

Sorry for any confusion.

**Comment by andersevenrud** [comment]
>  is not an option with that route, since it is not one of the assets that webpack is copying to the dist folder.

I assume that Webpack's compiler asset output API thingy don't support absolute paths.

But can't you:

1. Provide a path to the plugin that points to the metadata file (maybe there's a way to detect that)
2. Put the found assets into some kind of state of the plugin
3. When webpack is done, trigger on a "done" event and place the state above into the given metadata file path

**Comment by ajmeese7** [comment]
I haven't tested with the `dev-meta` repo yet since I'm currently developing on my fork, but I was able to get that working in a single package via a relative path, so it *should* work when abstracted out.

That leads to two questions:
1. How would I tap into that metadata in the `icon` function, given access to the `path` and the `theme` as there is currently?
2. Do you still want the default behavior to point to a `.png` even if it is known to not exist, or do you want `null` returned as it is in the `soundResource` function?

**Comment by ajmeese7** [comment]
Changes implemented as requested, I'll open the PR on `dev-meta` so you can test that things work on your end.

File: src/utils/desktop.js [hunk 2]
```diff
@@ -147,9 +148,13 @@ export const resourceResolver = (core) => {
 
   const soundsEnabled = () => !!getSoundThemeName();
 
-  const icon = path => {
+  const icon = (name) => {
     const theme = getThemeName('icons');
-    return core.url(`icons/${theme}/icons/${path}`); // FIXME: Use metadata ?
+    const extension = core.make('osjs/packages')
```
Comment (andersevenrud) [inline src/utils/desktop.js:153]: A few things here:

1. Probably better to hoist the service provider destructor to the top of the scope:
```js
export const resourceResolver = (core) => {
  const {getMetadataFromName} = core.make('osjs/packages');
```
2. This does not cover cases of invalid themes, nor if if the `icons` property is missing (make sure to create tests for this)
```js
const metadata = getMetadataFromName(theme)
const iconDefinitions = metadata ? metadata.icons || {} : {}
const extension = iconDefinitions[name] || 'png'
```

(I'm not sure if the current Babel setup takes care of optionals... above could be written as `iconDefinitions = metadata?.icons || {}` if that was the case. Gonna look into that).

File: index.d.ts [hunk 3]
```diff
@@ -1552,7 +1552,7 @@ declare class Tray {
 
 	/**
 	 */
-  has(key): boolean;
+	has(key): boolean;
```
Comment (andersevenrud) [inline index.d.ts:1555]: What's up with these indentation changes in this file ?

File: src/packages.js [hunk 4]
```diff
@@ -488,4 +488,13 @@ export default class Packages {
   running() {
     return this._running;
   }
+
+  /**
+   * Gets the package metadata for a given package name
+   * @param {string} name
+   * @returns {PackageMetadata}
+   */
+  getMetadataFromName(name) {
```
Comment (andersevenrud) [inline src/packages.js:497]: Return a copy here (as mentioned):

```
const found = this.metadata.find(pkg => pkg.name === name);
return found ? {...found} : null;
```

File: src/packages.js [hunk 5]
```diff
@@ -488,4 +488,13 @@ export default class Packages {
   running() {
     return this._running;
   }
+
+  /**
+   * Gets the package metadata for a given package name
+   * @param {string} name
+   * @returns {PackageMetadata}
```
Comment (andersevenrud) [inline src/packages.js:495]: This can return a null. So `@returns {PackageMetadata|null}`

File: src/utils/desktop.js [hunk 6]
```diff
@@ -147,9 +148,13 @@ export const resourceResolver = (core) => {
 
   const soundsEnabled = () => !!getSoundThemeName();
 
-  const icon = path => {
+  const icon = (name) => {
```
Comment (andersevenrud) [inline src/utils/desktop.js:151]: Maybe extensions should be removed here ? I imagine they are supplied in other places, so doing removal will ensure backward compatability.

You can probably add tests for this as well.

**Review by andersevenrud — CHANGES_REQUESTED** [review]

File: index.d.ts (hunk 3 above)
Comment (ajmeese7) [reply index.d.ts:1555]: 99% of the file was using tabs, so my local linter changed the remaining few lines with spaces to also have tabs. I can change all the lines to spaces if you would prefer, that would be no issue. Not sure why there was originally a mix, I do not believe that I introduced that.

File: src/utils/desktop.js (hunk 6 above)
Comment (ajmeese7) [reply src/utils/desktop.js:151]: Do you just want me to remove everything after/including the last period if it exists?

Comment (andersevenrud) [reply src/utils/desktop.js:151]: If the extension is not removed then this function would fail if you gave it `foo.png` as it would resolve to `foo.png.png`.

Comment (andersevenrud) [reply src/utils/desktop.js:151]: You can see this if you add some tests :)

Comment (ajmeese7) [reply src/utils/desktop.js:151]: I am aware of the double extension flaw; I was asking how you want that handled because of the edge case we discussed in `dev-meta` with the dotted file name issue. Removing everything after the last period could be problematic with an image name such as `windows.start.icon` for individuals who name things according to that format, so I'm unsure of the approach you would like to take to mitigate the issue.

I don't see the passing of an extension as much of an issue, I only noticed one place that used it and that was fixed in this PR (was for the taskbar search icon). So anyone who is updating will have the fix anyways. Regardless, I'm good with adding in a fix for extension compatibility if you have a suggestion of how to implement it.

And yes I will be adding tests 👍🏼

Comment (andersevenrud) [reply src/utils/desktop.js:151]: You can just use a list of known extensions and remove that. Chances are that someone naming stuff `icon.png.png` is slim :)

Comment (andersevenrud) [reply src/utils/desktop.js:151]: > Chances are that someone naming stuff

If this is the case then just slapping on some notices in the documentation would be the way to go. Maybe enforce some kind of pattern (which is why I went for GTK icon packs) ?

**Comment by ajmeese7** [comment]
I know these aren't the most comprehensive tests known to man, but for now I believe they get the job done. I would be happy to add more if there are additional aspects that you would like tested.

**Review by andersevenrud — CHANGES_REQUESTED** [review]
Looks like you reformatted the `.d.ts` file when changing it. Brought in quite a massive patch :)

File: __tests__/utils/desktop.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import {createInstance} from 'osjs';
+import {resourceResolver} from '../../src/utils/desktop';
+
+describe('#resourceResolver icon themes', () => {
+  let core;
+
+  beforeAll(() => createInstance().then((c) => (core = c)));
+  afterAll(() => core.destroy());
+
+  test('Should work with image extensions in the icon name', () => {
```
Comment (andersevenrud) [inline __tests__/utils/desktop.js:10]: This should also test for no extensions, filenames with dots in them, and double extensions.

**Review by andersevenrud — CHANGES_REQUESTED** [review]

**Comment by andersevenrud** [comment]
> forgot to run the tests to make sure there were no linter copy errors

Probably should add Husky or something like that to the core repos to prevent this sort of stuff.

File: __tests__/utils/desktop.js (hunk 7 above)
Comment (ajmeese7) [reply __tests__/utils/desktop.js:10]: > double extensions

To confirm your desired behavior, would you want `icon-name.gif.png` to become `icon-name` or `icon-name.gif`?

Comment (andersevenrud) [reply __tests__/utils/desktop.js:10]: > To confirm your desired behavior

The *required* behavior :)


Since this introduces extension resolution, all input has to be stripped of extensions. Preferably a set of known ones (as mentioned in an earlier discussion). The easiest way is probably just to use a regex that removes any supported extensions and the preceding dot.

So:

* `icon-name.png` -> `icon-name`
* `icon-name.foo` -> `icon-name.foo`
* `icon-name.png.png` -> `icon-name.png` (if someone does this, then it's on them imo. super edge-case)
* `icon-name.foo.png` -< `icon-name.foo`
* and so on

Comment (ajmeese7) [reply __tests__/utils/desktop.js:10]: Alright that's what I already wrote the tests for, perfect. Will make a commit soon 👍🏼

**Review by andersevenrud — CHANGES_REQUESTED** [review]
On the surface this seems fine now, with the exception of the `.d.ts` formatting that needs to be reverted.

**Comment by ajmeese7** [comment]
Alright now I'm confused, do you want a mix of tabs and spaces in the file ? In [this comment](https://github.com/os-js/osjs-client/pull/191#discussion_r963140143) you pointed out the indentation changes made by the linter, so in [this commit](https://github.com/os-js/osjs-client/pull/191/commits/1255f6a1f35b12ada9c602a4550eb4092a735a6e) I standardized the file to the space indentation that you normally use.

What do you want done to the file, do you want it left as spaces, changed to tabs, or reverted back to a random mix?

**Comment by andersevenrud** [comment]
Totally missed that comment.

Ideally, just add the updates to `.d.ts` then take care of formatting in a separate PR so that this goes out of context and is blown up with 2000+ lines of change :sweat_smile:

**Comment by andersevenrud** [comment]
Oh, I see you did a merge, and not a rebase. So this won't really work as expected... but a squash will brush all the stuff under a rug :D

With a rebase you would be able to "erase" the changes (history) from the bottom up, but a merge effectively does nothing to the history as it's on top.

**Comment by andersevenrud** [comment]
> With a rebase you would be able to "erase" the changes

Assuming you're in your branch you could do something like this:

```
# Update your master branch
git fetch origin master:master

# Start rebasing
git rebase master

# Now you'd hit a collision because there was changes to the type defs on both sides, so to resolve
# reset to the master file
git checkout --ours index.d.ts

# Now you can add your changes to the file, then
git add index.d.ts

# Finish up
git rebase --continue
```

Voila. The commit that had all of the changes to the file now only has the ones you initially was going to add.

And it can be force-pushed to the repo and the PR will reflect this :raised_hands:

**Review by andersevenrud — APPROVED** [review]
Code looks fine.

A little bit hesitant merging this without being able to test this personally. I'll see if I can't get it going on codesandbox or something so I can give it a merge sooner rather than later (I might not be home until Sunday)

File: __tests__/utils/desktop.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
+import {createInstance} from 'osjs';
+import {resourceResolver} from '../../src/utils/desktop';
+
+describe('#resourceResolver icon themes', () => {
```
Comment (andersevenrud) [inline __tests__/utils/desktop.js:4]: I think this can be condensed a bit. Instead of having all of these cases, maybe we can just compare arrays ?

```js
test('Should resolve correct paths', () => {
  const input = [
    'test.png',
    // and so on...
  ]

  const output = input.map(resolver.icon)

  expect(output).toEqual([
    '/icons/GnomeIcons/icons/test.png',
   // and so on...
  ])
})

```

Not a request, but something to look at it you want to.

**Comment by ajmeese7** [comment]
@andersevenrud reminder to check this out locally when you get a chance, no rush :)

**Comment by andersevenrud** [comment]
Sorry about the delay. Life got in the way :/ I'll get this merged this weekend.

**Comment by andersevenrud** [comment]
Gave it a go and it seems to be working fine :)

**Comment by andersevenrud** [comment]
This and all your other merged stuff was released under `3.9.0`.
